src/pushsource/_impl/backend/koji_source.py

```diff
@@ -509,6 +521,127 @@ def _push_items_from_container_build(self, nvr, meta):
 
         return out
 
+    def _push_items_from_vmi_build(self, nvr, meta):
+        LOG.debug("Looking for virtual machine images on %s, %s", nvr, meta)
+
+        if not meta:
+            message = "Virtual machine image build not found in koji: %s" % nvr
+            LOG.debug(message)
+            raise ValueError(message)
+
+        # VMI has the {'typeinfo': {'image':{}}} on extra
+        extra = meta.get("extra") or {}
+        typeinfo = extra.get("typeinfo") or {}
+        image = typeinfo.get("image")
+
+        if image is None:
+            message = "Build %s not recognized as a virtual machine image build" % nvr
+            LOG.debug(message)
+            raise ValueError(message)
+
+        build_id = meta["id"]
+
+        archives = self._get_archives(build_id)
+
+        # Supported types of VMI archives
+        ami_types = ["raw", "raw-xz"]
+        azure_types = ["vhd", "vhd-compressed"]
+        cloud_types = {}
+
+        for k in ami_types:
+            cloud_types.setdefault(k, AmiPushItem)
+
+        for k in azure_types:
+            cloud_types.setdefault(k, VHDPushItem)
+
+        # Collect all image archives which are not containers (tar)
+        # FIXME: Need to verify whether it's better to limit only to
+        # known extensions (like VHD, AMI) or not
+        vmi_archives = [
+            elem
+            for elem in archives
+            if elem.get("btype") == "image" and elem.get("type_name") != "tar"
+        ]
```

Comment:
@Tojaj @lslebodn @jajreidy in your opinion does it make sense for `KojiSource` to filter the image types only for VHD / AMI  or should it bring all image archives (which are not containers) whatever type it is?

Reply:
Do you know what image type that's going to be used for GCP? I know for amazon AMI can be a few file types.

Reply:
No idea... But yes, the disadvantage of filtering here is that we'll need to always keep the filter up to date...

Reply:
I don't think you could sensibly make this handle all types of 'image' archives.

I can't find an example build right now, but I'm pretty sure there are some builds having 'image' archives completely unrelated to container images and VM images.  Even if I'm wrong and it'd work today, it wouldn't be forwards-compatible in the case where new builds start to appear with types of 'image' not supported by this code.

Reply:
@rohanpm I've changed it to consider whether `docker` is present in extra, as well as verifying whether `container_koji_task_id` is not set, as suggested by @lslebodn .

PTAL again